### ISSUE TITLE
How to inject custom javascript to new windows requested by window.open?

### ISSUE THREAD

**Issue by ztanczos** [MicrosoftEdge/WebView2Feedback#2491]
Hi,

I'm trying to ensure that custom JavaScript is always injected to new windows which are requested by e.g.: window.open. For this I'm subscribing to NewWindowRequested event handler the following way:
```cs
        public async Task InitializeAsync()
        {
            await webView.EnsureCoreWebView2Async();
            webView.CoreWebView2.NewWindowRequested += CoreWebView2_NewWindowRequested;
        }

        private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {
            var deferral = e.GetDeferral();
            Window window = new Window();
            var newWebView = new WebView2();
            window.Content = newWebView;
            window.Show();
            await newWebView.EnsureCoreWebView2Async();
            await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("alert('hello')");

            e.NewWindow = newWebView.CoreWebView2;
            e.Handled = true;
            deferral.Complete();
        }
````
What I'm experiencing is that the new window pops up fine but the custom script is never executed.

How can I ensure that custom JavaScript is always executed when a new window is requested by window.open *before* the window object is returned to the JavaScript caller?

Thank you,
   Zoltan

[AB#43367417](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/43367417)

**Comment by vbryh-msft** [comment]
Hi @ztanczos , in order for JS injection to take effect in a new window, it should be called after `e.NewWindow = newWebView.CoreWebView2;`. [Here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2newwindowrequestedeventargs?view=webview2-1.0.1210.39#put_newwindow) is some doc.

**Comment by vbryh-msft** [comment]
@psmulovics, sorry for confusion. We have further `The methods which should affect the new web contents like AddScriptToExecuteOnDocumentCreated and add_WebResourceRequested have to be called after setting NewWindow.` - we will see how to make it more clear.

**Comment by pontusn** [comment]
Most interesting information, this enabled me to remove an ugly workaround where our integration actually had to reload ininitially loaded content for window.open to make all hooks and injected scripts work.

**Comment by ztanczos** [comment]
Hi @vbryh-msft,
Thank you for your answer but unfortunately it still doesn't work for me, at least not reliably. This is the full source code for the `NewWindowRequested` event handler:

```csharp
        private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {
            try
            {
                var deferral = e.GetDeferral();
                Window window = new Window();
                var newWebView = new WebView2();

                window.Content = newWebView;
                window.Show();

                await newWebView.EnsureCoreWebView2Async();
                e.NewWindow = newWebView.CoreWebView2;
                
                await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("alert('hello')");
                e.Handled = true;
                deferral.Complete();
            }
            catch (Exception ex)
            {

            }
        }
```

and this is the code which triggers the new window request:
```html
<button onclick="window.open('https://www.google.com')">click me</button>
```

With this event handler most of the time the popup window is just blank (i.e.: not even google.com is displayed), but sometimes - mostly when I put a breakpoint at `AddScriptToExecuteOnDocumentCreatedAsync` I see the page properly displayed and my 'hello' alert message as well.

Also if I put a `Thread.Sleep(3000)` before `AddScriptToExecuteOnDocumentCreatedAsync` it starts to work more often.

Am I doing something wrong or there's a synchronization issue in WV2?

**Comment by psmulovics** [comment]
tagging @liminzhu for tracking

**Comment by pontusn** [comment]
For our Win32 integration I avoid extensive processing in the event handler. Instead I get the deferral and use PostMessage to create window and integrate with new WebView2 instance from the message pump.

**Comment by ztanczos** [comment]
Hi @pontusn,
I'm not sure how would that be possible with the C# .NET interface: my understanding is that I'd need to set eventArgs.NewWindow to the new CoreWebView2 instance in the event handler and creating that instance can only be done asynchronously and it is already a relatively heavy operation.

Hi @vbryh-msft,
Just wondering if you had a chance to check my last message.

**Comment by pontusn** [comment]
Use the deferral and avoid lengthly processing in the event handler.

**Comment by vbryh-msft** [comment]
Hi @ztanczos  - I was able to repro the issue in .Net, but was unable to quickly solve it - created a bug for somebody to have more time to look into it.

**Comment by vbryh-msft** [comment]
@ztanczos  - I have tried your sample one more time - do you need to have `alert` in AddScriptToExecuteOnDocumentCreatedAsync? Could you try `console.log` instead and see how it behaves for you?

**Comment by ztanczos** [comment]
@vbryh-msft - I don't need `alert`, let me try to explain what I'm looking for: we have custom Javascript libraries which provide all sorts of functionalities for the hosted web applications. These libraries should be already present when loading a page. This works fine for the 'main' window but if a web application tries to open a new window by using `window.open` we'd like to intercept that call and inject the same set of JS libs _before_ the `window` object is available to the caller.

Consider this sample HTML:
```html
<html>
<head>
<title>test</title>
<script type="text/javascript">
	var openedWindow;
	
	console.log(window['TESTDATA'])
	
	function openWindow() {
	  openedWindow = window.open('https://www.google.com')
	  console.log('window opened')
	  console.log(openedWindow['TESTDATA'])
	}
</script>
</head>
<body>
<h1>hello world</h1>
<button onclick="openWindow()">click me</button>
</body>
</html>
```

And in .NET I'm trying to inject the following JS:
```cs
        public async Task InitializeAsync()
        {
            await webView.EnsureCoreWebView2Async();
            webView.CoreWebView2.NewWindowRequested += CoreWebView2_NewWindowRequested;
            await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window['TESTDATA'] = 'this works'");
        }
```
```cs
        private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {
            try
            {
                _deferral = e.GetDeferral();
                Window window = new Window();
                var newWebView = new WebView2();

                window.Content = newWebView;
                window.Show();

                await newWebView.EnsureCoreWebView2Async();
                e.NewWindow = newWebView.CoreWebView2;

                await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window['TESTDATA'] = 'this doesn't work'");
                e.Handled = true;
                _deferral.Complete();
            }
            catch (Exception ex)
            {
                Debug.Fail(ex.ToString());
            }
        }
```

When I click on the button `undefined` is logged to the console instead of _this doesn't work_.

Not sure if this makes sense, let me know if you have further questions.
Thank you

**Comment by vbryh-msft** [comment]
@ztanczos There are could be two issues with your example:
1. Apostrophe in `doesn't` throws `Uncaught SyntaxError: Unexpected identifier` - could you try without it, like just `does`
2. Parent window from where you are trying to access `console.log(openedWindow['TESTDATA'])` matters - could you try to open google.com from google.com? Hope it makes sense. Or just check `window['TESTDATA']`  in DevTools(right-click -> Inspect) of the new window?
![Screenshot_doeswork](https://user-images.githubusercontent.com/72414887/174853526-e7b0e595-4c8d-4b97-80bf-2e28f5671e16.png)

**Comment by ztanczos** [comment]
@vbryh-msft thanks, I didn't notice the apostrophe.. 🤦 
I changed the test URLs so there's no cross-origin issue and now I see the same behavior, i.e.: `undefined` is logged when the window is opened but later I see `TESTDATA` populated.
I'll check whether this behavior would work for us.

**Comment by ztanczos** [comment]
What exactly is the purpose of calling `deferral.Complete()`? I commented out that line so that I never call `.Complete()` and I don't see any change in behavior.

**Comment by vbryh-msft** [comment]
Hi @ztanczos  - you should use `GetDeferral` if you want to defer event completion at a later time. [Here is an example](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/WebView2WpfBrowser/MainWindow.xaml.cs#:~:text=CoreWebView2Deferral%20deferral%20%3D,.Complete()%3B) how it can be used - you can search for more examples in that file. We suggest to always complete deferral if it was taken in order for event to be in proper state and be able to complete successfully.

Is injecting JS script works in new window as expected?

**Comment by plantree1995** [comment]
Hi @ztanczos, I have tried some tests based on `WPF_GettingStarted` and it seems fine. My code is as follows:
```C#
public async Task InitializeAsync()
{
	await webView.EnsureCoreWebView2Async();
	webView.CoreWebView2.NewWindowRequested += CoreWebView2_NewWindowRequested;
	//await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window['TESTDATA'] = 'this works'");
}

private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
{
	var _deferral = e.GetDeferral();
	Window window = new Window();
	var newWebView = new WebView2();

	window.Content = newWebView;
	window.Show();

	await newWebView.EnsureCoreWebView2Async();
	e.NewWindow = newWebView.CoreWebView2;

	await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window['TESTDATA'] = 'this does work'; alert('hello')");
	await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("function square(x) { return x * x; }");

	e.Handled = true;
	_deferral.Complete();
}
```
You will get the `alert` and the function works.

**Comment by ztanczos** [comment]
Hi @plantree1995 ,
Your code sample doesn't work for me: I don't see any alert, in fact the new window opens as completely blank.

**Comment by psmulovics** [comment]
> Hi @plantree1995 , Your code sample doesn't work for me: I don't see any alert, in fact the new window opens as completely blank.

Same for me.

**Comment by plantree1995** [comment]
Hi @ztanczos @psmulovics, I followed this [project](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/GettingStartedGuides/WPF_GettingStarted), and all source code is posted below:
```C#
namespace WPFSample
{
    /// <summary>
    /// Interaction logic for MainWindow.xaml
    /// </summary>
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            webView.NavigationStarting += EnsureHttps;
            InitializeAsync();

        }

        async void InitializeAsync()
        {
            await webView.EnsureCoreWebView2Async(null);
            webView.CoreWebView2.WebMessageReceived += UpdateAddressBar;
            webView.CoreWebView2.NewWindowRequested += CoreWebView2_NewWindowRequested;

            await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.chrome.webview.postMessage(window.document.URL);");
            await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.chrome.webview.addEventListener(\'message\', event => alert(event.data));");

        }

        void UpdateAddressBar(object sender, CoreWebView2WebMessageReceivedEventArgs args)
        {
            String uri = args.TryGetWebMessageAsString();
            addressBar.Text = uri;
            webView.CoreWebView2.PostWebMessageAsString(uri);
        }

        private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
        {
            var _deferral = e.GetDeferral();
            Window window = new Window();
            var newWebView = new WebView2();

            window.Content = newWebView;
            window.Show();

            await newWebView.EnsureCoreWebView2Async();
            e.NewWindow = newWebView.CoreWebView2;

            await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window['TESTDATA'] = 'this does work'; alert('hello')");
            await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("function square(x) { return x * x; }");

            e.Handled = true;
            _deferral.Complete();
        }

        void EnsureHttps(object sender, CoreWebView2NavigationStartingEventArgs args)
        {
            String uri = args.Uri;
            if (!uri.StartsWith("https://"))
            {
                webView.CoreWebView2.ExecuteScriptAsync($"alert('{uri} is not safe, try an https link')");
                args.Cancel = true;
            }
        }

        private void ButtonGo_Click(object sender, RoutedEventArgs e)
        {
            if (webView != null && webView.CoreWebView2 != null)
            {
                webView.CoreWebView2.Navigate(addressBar.Text);
            }
        }

    }

}

```
Compile and run it in debug mode, open any links in new window, I would get the `alert` and `function square()`. If you have any other questions, please let me know.

**Comment by ztanczos** [comment]
Hi @plantree1995,
This sample is slightly different from what we're trying to achieve. We're trying to open a new window using Javascript's `window.open()` this way:

```
<html>
<head>
<title>test</title>
<script type="text/javascript">
	var openedWindow;
	
	console.log(window['TESTDATA'])
	
	function openWindow() {
	  openedWindow = window.open('http://mytesturl/wv2test.html')
	  console.log(openedWindow['TESTDATA']);
	}
</script>
</head>
<body>
<h1>hello world</h1>
<button onclick="openWindow()">click me</button>
</body>
</html>
```

When I open this HTML with the sample code from WPF_GettingStarted I see the main page rendered correctly but when I click on the `click me` button I only get a blank window and no `alert` pop-up + `undefined` is logged to the console indicating that `TESTDATA` is not available on `openedWindow`.

**Comment by plantree1995** [comment]
Hi @ztanczos, I tried this way.
```C#
      private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
      {
          try
          {
              var _deferral = e.GetDeferral();
              Window window = new Window();
              var newWebView = new WebView2();

              window.Content = newWebView;
              window.Show();

              await newWebView.EnsureCoreWebView2Async();
              e.NewWindow = newWebView.CoreWebView2;

              await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("function square(x) { return x * x; }; function cube(x) { return x * x * x; }");


              e.Handled = true;
              _deferral.Complete();
          }
          catch (Exception )
          {

          }
      }
```
and the html just like this:
```html
<html>

<head>
  <title>test</title>
  <script type="text/javascript">
    var openedWindow;

    console.log(window['TESTDATA'])

    function openWindow() {
      openedWindow = window.open('http://bing.com', 'hello')
    }
  </script>
</head>

<body>
  <h1>hello world</h1>
  <button onclick="openWindow()">click me</button>
</body>

</html>
```
When you click on the button to open a new window, you could use `function square()` and `function cube()` in the **Console**.

**Comment by ztanczos** [comment]
Hi @plantree1995,
Thanks for looking into this but I know that _eventually_ the injected Javascript code will be there but I have no control over when and I'd need to ensure that the returned Window object already has the injected JS functionality **right after** the `window.open()` call.

Our use-case: we're hosting web applications in our container which provides rich functionality for these web applications via JS + interop objects. This functionality is required in windows opened by these web applications too _without timing issues_. In the past we could achieve the same with CEF + CefSharp by using V8Extensions but now we're struggling with WebView2.

**Comment by ztanczos** [comment]
The other way around. E.g.: let's say we inject something into the parent window so that `window.MyCustomCode` is accessible & works. Right after `openedWindow = window.open(...)` I'd like to ensure that `openedWindow.MyCustomCode` also accessible & works.
In reality it is a bit more complex than that (multiple JS libraries with multiple properties, classes, etc.) but this is the gist of it.

**Comment by plantree1995** [comment]
Hi @ztanczos,
I almost understand what you want. It seems that you can read this [docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/open) to known more about what effects will `Window.open()` have.
> Note that remote URLs won't load immediately. When window.open() returns, the window always contains about:blank. The actual fetching of the URL is deferred and starts after the current script block finishes executing. The window creation and the loading of the referenced resource are done asynchronously

It's not synchronous, so the output `undefined` is as expected.

**Comment by plantree1995** [comment]
Maybe we can better illustrate this issue: *the function of open returns although the resource has not yet been loaded completely*.
So perhaps you can use in this way:
```js
function openWindow() {
  openedWindow = await window.open('http://127.0.0.1:8010/foo.html');
  openedWindow.onload = function () {
    console.log(openedWindow['TESTDATA']);
  };
}
```

**Comment by plantree1995** [comment]
Hi @ztanczos, have you already solved this problem? If so, I will close this issue.

**Comment by ztanczos** [comment]
Hi @plantree1995,
Sorry for the late response, I was on holiday. I'll still need a few days, please don't close this issue yet.

**Comment by plantree1995** [comment]
Hi @ztanczos,
If you come cross any problems, please let me know. Perhaps we could tackle them together.

**Comment by ztanczos** [comment]
Hi @plantree1995,
Please bear with me, our actual problem spans across multiple applications and libraries, some of which are open-source, some are not so it is difficult to describe the problem without sharing non-public information.

With Cef + CefSharp we "injected" (or rather registered) the custom Javascript libraries by using V8Extensions (`CefSettings.RegisterExtension()` method). That way the custom Javascript is always _there_, even for newly opened windows, immediately after calling `window.open` (so it is never `undefined`).

Our framework calls `window.open` through desktopJS: https://github.com/morganstanley/desktopJS/blob/main/packages/desktopjs/src/Default/default.ts#L299 - we might be able to tweak that for our needs.

You wrote that injecting JS is not synchronous - what would be your suggestion to check when exactly a custom JS script is injected & ready to be consumed?

**Comment by plantree1995** [comment]
Hi @ztanczos, follow this link, [Window.open()](https://developer.mozilla.org/en-US/docs/Web/API/Window/open#description), we could know that:

> The window creation and the loading of the referenced resource are done asynchronously.

In a previous issue, I have this suggestion.

> Maybe we can better illustrate this issue: _the function of open returns although the resource has not yet been loaded completely_. So perhaps you can use in this way:
> 
> ```js
> function openWindow() {
>   openedWindow = window.open('http://127.0.0.1:8010/foo.html'); // remove the await
>   openedWindow.onload = function () {
>     console.log(openedWindow['TESTDATA']);
>   };
> }
> ```

I don't know have you ever tried in this way. You can use your custom JS scripts freely after the `load` event is fired. The reference is below: https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event#syntax .

**Comment by ibebbs** [comment]
Hi @ztanczos / @vbryh-msft / @plantree1995 ,

I also believe there is a race condition between setting `CoreWebView2NewWindowRequestedEventArgs.NewWindow` and having scripts loaded (via `CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync`) prior to navigation / document loading.

Based on the above, I have put together a full reproduction of the issue in [this repo](https://github.com/ibebbs/ScriptVsNewWindow) along with steps for (and recordings of) expected and actual behavior.

I would very much appreciate if people could validate this issue and/or provide a work around.

FWIW, my hunch is that WebView2 is scheduling a navigation to the destination URL on the UI thread as soon as `CoreWebView2NewWindowRequestedEventArgs.NewWindow` is assigned. This causes the navigation to take place as soon as the currently executing code yields the UI thread (via a `Task.Delay` or - I expect - somewhere in `CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync`), potentially before the scripts have been added. If this is the case, a simple fix for this issue would be to wait for the deferral to be completed prior to scheduling the navigation.

**Comment by ibebbs** [comment]
I have pushed a workaround for this issue to [this branch](https://github.com/ibebbs/ScriptVsNewWindow/tree/cancel_initial_navigation).

Short answer: Use `CoreWebView2.NavigationStarting` to cancel any navigations that occur prior to `CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync` completing then schedule the desired navigation manually.

Long answer: The issue was that setting CoreWebView2NewWindowRequestedEventArgs.NewWindow caused WebView2 to immediately (regardless of whether a deferral was held or not) schedule a navigation to CoreWebView2NewWindowRequestedEventArgs.Uri on the UI thread. Subsequent calls to CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync yielded the UI thread establishing a race-condition condition between the navigation and the script loading (hence why there were different outcomes based on timing / CPU load / etc).

Regarding a "fix" for WebView2 to be discussed with MS, I would suggest:
1. __Critial__: Setting NewWindow should not schedule a navigation until any held deferral has been completed.
2. _Nice to have_: CoreWebView2NewWindowRequested should provide a `Navigate` boolean which allows you to instruct WebView2 whether or not to schedule a navigation.

**Comment by ztanczos** [comment]
Hi @ibebbs,
Our use-case is a bit special because we're in fact opening a new *blank* window, i.e.: about:blank, and once opened we're directly manipulating and projecting DOM elements to the new window. 

Because a blank window is special we ended up using `ExecuteScriptAsync` instead of `AddScriptToExecuteOnDocumentCreatedAsync` and also we triggered an "extra" navigate on the newly opened window and then we're essentially waiting for the injected scripts / interop objects to become available on the Javascript side. We're using desktopJS and [here](https://github.com/morganstanley/desktopJS/blob/main/packages/desktopjs/src/Default/default.ts#L331) instead of returning `newWindow` we return a Promise.

This _works_ although we'd really appreciate a more straightforward way of achieving the same.

**Comment by vbryh-msft** [comment]
Hi @ibebbs  - thank you for providing the repro repo. I have checked it - it should work if you will remove `alert` from the script and put `console.log` instead. Could you please try that? Do you really need to have alert in script or it was just for demo purpose? As mentioned before in general flow with adding script for new window should work unless you have some corner case also - then we would like to learn about what are you trying to achieve.

**Comment by RendijsSmukulis** [comment]
@vbryh-msft I've just tried @ibebbs repro with alerts replaced with `console.log`s, with the same outcome 🤔

**Comment by vbryh-msft** [comment]
@RendijsSmukulis could you please check this comment https://github.com/MicrosoftEdge/WebView2Feedback/issues/2491#issuecomment-1211604701 - it summarizes the issue, thanks.

**Comment by vbryh-msft** [comment]
@RendijsSmukulis @ibebbs also to reiterate - this is correct order of calls, which should work, do not set NewWindow after adding the script:
```
	window.Content = newWebView;
	window.Show();

	await newWebView.EnsureCoreWebView2Async();
	e.NewWindow = newWebView.CoreWebView2;

	await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("console.log('hello')");
	e.Handled = true;
	_deferral.Complete();
```

**Comment by ibebbs** [comment]
Thanks for the looking into this @vbryh-msft.

I do not believe the use of `alert` is the issue here. While using `console.log` may have appeared to resolve the issue for you, as demonstrated by @RendijsSmukulis, it does not work reliably due to what appears to be an underlying race condition (navigation occuring as soon as the UI thread is relinquished).

To demonstrate this I have added logging to the [master branch of the reproduction repository](https://github.com/ibebbs/ScriptVsNewWindow) and changed the script to use `console.log` rather than `alert` per your suggestion.

If you run this app, ensure "Set NewWindow" is checked and "Set Source" is not, then click the "Open Window From Target..." link, you we see the following log:

```
[02:56:41.3058] Assigning NewWindow
[02:56:41.3134] Assigned NewWindow
[02:56:41.3135] Start Loading Scripts
[02:56:41.3623] NavigationStarting - 'about:blank'
[02:56:41.6923] ContentLoading - 'https://ian.bebbs.co.uk/'
[02:56:41.6984] Completed Loading Scripts
```

As you can see, content loading has been started prior to script loading being complete. __This is the core of the issue__. However, this is exacerbated by _anything that might cause the UI thread to be relinquished_. For example, if you clear the log and set "Delay" to "2000" then click the "Open Window From Target..." link again, you will see the following log:

```
[03:08:37.2107] Assigning NewWindow
[03:08:37.2179] Assigned NewWindow
[03:08:37.2816] NavigationStarting - 'about:blank'
[03:08:37.5168] ContentLoading - 'https://ian.bebbs.co.uk/'
[03:08:39.2501] Start Loading Scripts
[03:08:39.2747] Completed Loading Scripts
```

As you can see, content has loaded before the scripts have even started loading.

I believe this demonstrates the core of the issue and why navigation should not be scheduled prior to the deferral being completed.

Thanks again for you help here.

**Comment by ibebbs** [comment]
No problem @plantree, glad to be of help.

Could you provide more information about your `PostMessage()` workaround? The workaround we're currently employing - cancelling navigations until script loading is complete - has a serious issue (navigation via form POST no longer works correctly) and we will need to find a better solution if MS are not able to provide a fix for this issue in the short-term.

Thanks!

**Comment by plantree** [comment]
Hi @ibebbs,

I prepared a demo project to show the workaround, which could be found [here](https://github.com/plantree/WebView2Samples/commit/015042077b7c3465d1edbe78b133f50116aa0ce9#diff-81720895a99cdddee93ccba366c91f50cf0a04a875b4ff4c2b7b6bc978467e78). If you have any questions, please let me know.

Besides, your suggestions are very helpful, and we are working on solving the root cause. 

Thanks for you advice.

**Comment by ibebbs** [comment]
Thanks for the demo project @plantree 

I see what you're doing and how it might be a viable workaround for some. Unfortunately I don't think we'll be able to use this approach as:
1. We display external web content so can't (easily) ensure clicking a link would invoke the `openWindow` function; and
2. It doesn't look like it'll resolve the issue with navigation via form POST.

Do you have an ETA on when a fix for the root cause might be available?

Thanks again for all your help.

**Comment by plantree** [comment]
Hi @ibebbs,

We do not have ETA currently, but I'll synchronize with you when we have.

As for you question, if it is possible, could you provide **a demo project** that we could try to solve together?

**Comment by ibebbs** [comment]
Sorry for the delay in getting back to you @plantree.

Unfortunately we don't have a demo project but [the repo provided above](https://github.com/ibebbs/ScriptVsNewWindow) should be sufficient to address the issue.

Thanks!

**Comment by plantree** [comment]
Hi @ibebbs,

Thanks for the demo project you provide. I'll try to repoduce your issue locally and synchronize with you later.

**Comment by ibebbs** [comment]
Hey @plantree, any update on this issue or an ETA for a fix?

**Comment by monica-ch** [comment]
Hi everyone, we fixed the issue that allows app to inject initial scripts by calling AddScriptToExecuteOnDocumentCreated before new window is created in canary version 114.0.1794 and above.

```
webView.CoreWebView2.NewWindowRequested += (sender, args) =>
{
      string id = await newWebview.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("console.log('Hello');");
      args.NewWindow = newWebView.CoreWebView2;
      args.Handled = true;
};
```

Please validate and let us know if you have any questions.

cc: @plantree @vbryh-msft

**Comment by plantree** [comment]
Hi @ztanczos, @ibebbs and @q71114,

You can test the new feature after switching to the preview channel, and the reference documentation could be found here: https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel. 

Any feedbacks will be welcome. Thanks.

**Comment by q71114** [comment]
Hi @monica-ch  and @plantree,

Thanks for the fix! I [forked the test project](https://github.com/q71114/ScriptVsNewWindow) created by @ibebbs and updated it to work with the latest canary (`114.0.1803.0`). Here are my test results:

- 🟢 New window created with GET navigation works well.
![image](https://user-images.githubusercontent.com/8770144/233052367-a3bbd510-5a0e-4bff-97f4-ba54443f6b17.png)

- 🟡 New window without navigation using JS manipulation works but the script is injected after the document is being altered by JS. 
![image](https://user-images.githubusercontent.com/8770144/233052180-61d0088a-5c83-4b56-92fd-39b771c8cc1e.png)

- 🔴 New window with POST navigation doesn't work
![image](https://user-images.githubusercontent.com/8770144/233052076-47bfa29a-64b2-49b6-b352-9b0bdd2ceb36.png)

Let me know if I'm missing something with the last test case.

**Comment by plantree** [comment]
Hi @q71114,

**Thanks for your quick feedbacks and well prepared example.**

- For the second test, it seems that the behavior is as expected, and you could find reference here: https://developer.mozilla.org/en-US/docs/Web/API/Window/open#description.

  > Note that remote URLs won't load immediately. When window.open() returns, the window always contains about:blank. The actual fetching of the URL is deferred and starts after the current script block finishes executing. The window creation and the loading of the referenced resource are done asynchronously.

- For the third, I have reproduced locally, and if you have more try, the same behavior occurs in `<a target="_blank">`. I am still  not sure if this is a bug, but I'll keep track of this thing and sync with you when there is progress.

If you have any other questions, feel free to let me know. Thanks.

**Comment by plantree** [comment]
Hi @q71114,

Add `form.setAttribute('rel', 'opener')`, and you will get the expected.

The reasons behind it are a bit complicated, and you could refer this link from MDN: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/noopener. In general, browser will restrict some behaviors of the new window opened with `noopener`.

Hope to solve your issues.

**Comment by q71114** [comment]
Hi @plantree,

Thanks for the pointers and the quick response. I updated [the test project](https://github.com/q71114/ScriptVsNewWindow) with your suggestion and it seems to work properly with POST navigation as well.
![image](https://user-images.githubusercontent.com/8770144/233593242-9c6e4a50-daa1-4e89-98e9-0d51eea768a7.png)

We will test this fix further with our other test cases - non script injection related - and get back to you with the results.

**Comment by gkerenyi** [comment]
Hi @plantree,

In our use case we need to inject a script to every opened page, but we can't do that in the new window opened by `<a target="_blank">' or `<form target="_blank">` if the `rel="opener"` attribute of the link/form is not set (which we don't control).
I checked the linked MDN page, but it does not explain restrictions on loading scripts, only that scripts would not have access to the opening document. From that perspective, I don't see the difference between a script loaded by the target page and one injected to it.
Could you please elaborate on (or point me to an explanation) why script injection is not allowed in this case? Thanks!

**Comment by plantree** [comment]
Hi @gkerenyi,

Unfortunately, if we don't set `rel="opener"` to `<a>` or `<form>`, perhaps we could not inject our custom scripts successfully.

Follow this [page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#security_and_privacy), in safefy considerations, setting `target="_blank"` implicitly provides the same protection as setting `rel="noopener"`. 

> The `noopener` keyword for the [rel](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel) attribute of the [a](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), [area](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/area), and [form](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) elements instructs the browser to navigate to the target resource **without granting the new browsing context access to the document that opened it**.

Relative link could be found here: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/noopener

Hope to solve your problem.

**Comment by gkerenyi** [comment]
Hi @plantree,

Thanks for getting back to me.

I have tested script injection with an `<a rel="noopener">` element (**without** setting `target="blank"`), and the injected script worked in that case.

Based on this, my understanding is that if `target="blank"` implicitly provides the same protection as `rel="noopener"`, script injection should be allowed in that case as well.

We'd like to maintain our request to allow script injection for all opened documents.

**Comment by plantree** [comment]
Hi @gkerenyi,

Thanks for your professional reply!

According to your description, the result should be as expected. According to the [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), using `<a>` without given `target` means default behavior, which is `_self`, and the current window will be reused. No new window will be opened, so in my story, `rel="noopener"` actually does nothing.

To help better understand your need, could you provide the scenario where you need to inject scripts to all opened window?

**Comment by gkerenyi** [comment]
Hi @plantree,

> To help better understand your need, could you provide the scenario where you need to inject scripts to all opened window?

We are building a generic browser where we inject a custom script to each page for various purposes (e.g. analyzing content, showing overlays on some elements, etc.). Script injection should work in all pages regardless of the way they were opened, otherwise the script's functionalities would depend on whether the page was opened directly or in a new tab/window.

> Unfortunately, if we don't set `rel="opener"` to `<a>` or `<form>`, perhaps we could not inject our custom scripts successfully.

With all due respect, this sounds like an assumption. I don't understand why script injection would not be allowed with `rel="noopener"`. The linked documentation states that `rel=noopener` means the scripts of the new window do not have access to the opening window. From this point of view, why would an injected script be different from a script that is loaded by the new page itself (which is allowed with `rel=noopener`)?

**Comment by plantree** [comment]
Hi @gkerenyi,

Thanks for your professional and patient analysis! As for your concern, it might be a problem, and I'll keep track of this and try to fix if it is possible. 

Any progress will be synchronized with you in time. If you have any other questions, please feel free to let me know.

**Comment by plantree** [comment]
Hi @gkerenyi,

After some investigations, I have found an easy way to make it work. If you're available, please have a try.

Do a navigate at the end of handling `NewWindowRequestEvent`, like this:

```c#
private async void CoreWebView2_NewWindowRequested(object sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs e)
{
    var deferral = e.GetDeferral();
    Window window = new Window();
    var newWebView = new WebView2();
    window.Content = newWebView;
    window.Show();
    await newWebView.EnsureCoreWebView2Async();
    await newWebView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("alert('hello')");
    newWebView.CoreWebView2.NavigationStarting += (s, ev) =>
    {
        Debug.WriteLine("-------------New WebView2 NavigationStarting: " + ev.Uri);
    };
    e.NewWindow = newWebView.CoreWebView2;
    e.Handled = true;
    deferral.Complete();
    // Will work in `noopener` mode
    newWebView.CoreWebView2.Navigate(e.Uri);
}
```

I will still keep focus on this and try to find an easier and more elegant way. Any feedbacks will be welcome. Thanks!

**Comment by gkerenyi** [comment]
Hi @plantree,

Thank you for your suggestion, but unfortunately it is not satisfactory, because it doesn't work with POST requests (the POST data is lost after re-navigating). In fact, this kind of re-navigation is the current workaround we are using to mitigate the original timing problem of script injection in the stable version of WebView2, and it is a major issue for us that POST requests in new windows are failing with this workaround.

To give you background, many sites (e.g., banking sites) open new windows using POST navigation. We need script injection to work in such a situation and we need the navigation to get the POST data through.

To summarize:
- Script injection should work in all pages regardless of the way they were opened.
- We can't use `Navigate()` but need to rely on the parent WebView to handle the navigation for the child WebView, otherwise POST navigation will lose the POST data.

Thank you for your continued focus on this problem!

**Comment by plantree** [comment]
Hi @gkerenyi,

Thanks for your patience and meticulous analysis! 

If you're available, could you provide a simple test demo, which has a complete coverage of the scenes you need. That might be very helpful while we are investigating this issue :) .

**Comment by gkerenyi** [comment]
Hi @plantree,

I forked a demo project previously used in this thread and simplified it to demonstrate the relevant test cases:
https://github.com/gkerenyi/ScriptVsNewWindow

Please see the README file for details on the actual and expected behaviors.

**Comment by plantree** [comment]
Hi @gkerenyi,

Thanks for your well-prepared demo project! 

I'll check that and any progress will be synchronized to you ASAP.

**Comment by MarkIngramUK** [comment]
@plantree , what's the status of this? Is there an ETA for a fix?

CC @nishitha-burman

**Comment by plantree** [comment]
Hi @MarkIngramUK,

Thanks for you attention! We have already prepared a fix solution, which is under review internally. I will try to get back with the ETA a few days later. 

Thanks for your patience.

**Comment by plantree** [comment]
Hi @MarkIngramUK,

Apologies for my delayed response. I needed to ensure that I was up-to-date with the latest developments. 

The fix patch has been included in the most recent Canary release, such as version `117.0.2024.0` or later. Feel free to give it a try, and your feedbacks would be greatly appreciated. Thanks!

cc. @novac42

**Comment by plantree** [comment]
Hi @psmulovics  and @ztanczos,

Thanks for your interest. Fortunately, **this alteration takes place during runtime, eliminating the need for a NuGet update.** Simply utilize the latest Canary version and verify its effects. Your feedbacks will be greatly appreciated. Thanks!

**Comment by gkerenyi** [comment]
Hi @plantree,

Thanks for the notification about the fix. I have tested the solution and so far it is looking good.
We will continue our testing and get back to you in case we find anything.

**Comment by plantree** [comment]
Hi @gkerenyi,

Thank you for sharing your feedback! Feel free to reach out whenever you have questions. Have a nice day.

**Comment by pontusn** [comment]
I've tried latest Canary 118.* and there are still something strange happening related to first navigation via window.open.

When we actually get correct script injection the mechanism behind WebMessage fails, breaking both our custom integrations and general support for COM.

**Comment by plantree** [comment]
Hi @pontusn,

Thanks for your feedback! To help better understand your issue, could you provide a simple demo for reproduction? Thanks.

**Comment by gkerenyi** [comment]
Hi @plantree & @victorhuangwq,

We found a timing problem similar to script injection, but this time with web resource request handling, and it can be reproduced both with the `WebResourceRequested` event and the Chrome DevTools Protocol `Fetch.requestPaused` event.

The essence of the issue is the following:
- Assigning a handler to these events before setting `CoreWebView2NewWindowRequestedEventArgs.NewWindow` has no effect (handlers are not called).
- Assigning a handler to these events after setting `CoreWebView2NewWindowRequestedEventArgs.NewWindow` results in a race condition, because navigation is already started in the background when setting `NewWindow`, so nothing guarantees that we already have a handler in place when the first web resource requests need to be processed.

I created a small repro app, please find it here with additional information in the readme: https://github.com/gkerenyi/CdpVsNewWindow

Either it should be possible to set up both types of event handlers before setting `NewWindow`, or navigation should not start until the deferral for `NewWindowRequested` is completed.

Since the `WebResourceRequested` event is lacking in functionality compared to CDP `Fetch`, we especially need the `Fetch` method to be guaranteed to be ready to process all web requests when starting navigation in a new window.

**Comment by victorhuangwq** [comment]
@yildirimcagri-msft could you take a look at this issue?

**Comment by victorhuangwq** [comment]
@yildirimcagri-msft looks like there's regression on something that was previously addressed - could you take a look?

**Comment by gkerenyi** [comment]
@pontusn As discussed before in relation to script injection, re-navigation is not an option, because it loses POST data.
The same is true for web request processing as for script injection: it should be possible to hook everything up before navigation is started in the new window by the opener window.

**Comment by pontusn** [comment]
Agree, but for our usage retry works.

**Comment by RendijsSmukulis** [comment]
@nishitha-burman as discussed, I've created a new issue (https://github.com/MicrosoftEdge/WebView2Feedback/issues/4181) as the discussion in this bug report has deviated from the original issue (script injection).

**Comment by champnic** [comment]
Apparently this original issue was fixed back in runtime versions 114+. Closing as completed.
